Framework/Algorithms/src/SofQWCentre.cpp

```diff
@@ -141,109 +131,97 @@ void SofQWCentre::exec() {
+    // does not hold. In order to
+    // be completely general we must calculate the momentum transfer by
+    // calculating the incident and final
+    // wave vectors and then use |q| = sqrt[(ki - kf)*(ki - kf)]
+
+    const auto &detIDs = inputWorkspace->getSpectrum(i).getDetectorIDs();
+    double numDets_d = static_cast<double>(detIDs.size());
+    const auto &Y = inputWorkspace->y(i);
+    const auto &E = inputWorkspace->e(i);
+    const auto &X = inputWorkspace->x(i);
+
+    // Loop over the detectors and for each bin calculate Q
+    for (const auto detID : detIDs) {
+      size_t idet;
+      try {
+        idet = detectorInfo.indexOf(detID);
+      } catch (std::out_of_range &) {
+        // Skip invalid detector IDs
+        numDets_d -= 1.0;
+        continue;
       }
-
-      const size_t numDets = detectors.size();
-      // cache to reduce number of static casts
-      const double numDets_d = static_cast<double>(numDets);
-      const auto &Y = inputWorkspace->y(i);
-      const auto &E = inputWorkspace->e(i);
-      const auto &X = inputWorkspace->x(i);
-
-      // Loop over the detectors and for each bin calculate Q
-      for (size_t idet = 0; idet < numDets; ++idet) {
-        IDetector_const_sptr det = detectors[idet];
-        // Calculate kf vector direction and then Q for each energy bin
-        V3D scatterDir = (det->getPos() - sample->getPos());
-        scatterDir.normalize();
-        for (size_t j = 0; j < numBins; ++j) {
-          const double deltaE = 0.5 * (X[j] + X[j + 1]);
-          // Compute ki and kf wave vectors and therefore q = ki - kf
-          double ei(0.0), ef(0.0);
-          if (emode == 1) {
-            ei = efixed;
-            ef = efixed - deltaE;
-            if (ef < 0) {
-              std::string mess =
-                  "Energy transfer requested in Direct mode exceeds incident "
-                  "energy.\n Found for det ID: " +
-                  std::to_string(idet) + " bin No " + std::to_string(j) +
-                  " with Ei=" + boost::lexical_cast<std::string>(efixed) +
-                  " and energy transfer: " +
-                  boost::lexical_cast<std::string>(deltaE);
-              throw std::runtime_error(mess);
-            }
-          } else {
-            ei = efixed + deltaE;
-            ef = efixed;
-            if (ef < 0) {
-              std::string mess =
-                  "Incident energy of a neutron is negative. Are you trying to "
-                  "process Direct data in Indirect mode?\n Found for det ID: " +
-                  std::to_string(idet) + " bin No " + std::to_string(j) +
-                  " with efied=" + boost::lexical_cast<std::string>(efixed) +
-                  " and energy transfer: " +
-                  boost::lexical_cast<std::string>(deltaE);
-              throw std::runtime_error(mess);
-            }
+      // Calculate kf vector direction and then Q for each energy bin
+      V3D scatterDir =
```

Comment:
Yes, that would make sense, I think. We can add this once the basics are more settled.